meshmode/array_context.py

```diff
@@ -270,184 +285,222 @@ def get_container_context_recursively(ary):
+                f"for {arg1_entry} in {arg1_container}")
+
+    elif isinstance(arg2, tuple):
+        arg2_entry, arg2_container = arg2
+        return (f"{op_str.format(arg1, arg2_entry)} "
+                f"for {arg2_entry} in {arg2_container}")
+    else:
+        return op_str.format(arg1, arg2)
+
+
+def with_container_arithmetic(
+        bcast_number=True, bcast_obj_array=None,
+        arithmetic=True, bitwise=False, shift=False,
+        eq_comparison=True, rel_comparison=None):
+    if bcast_obj_array is None:
+        raise TypeError("bcast_obj_array must be specified")
+    if rel_comparison is None:
+        raise TypeError("rel_comparison must be specified")
+
+    desired_op_classes = set()
+    if arithmetic:
+        desired_op_classes.add(_OpClass.ARITHMETIC)
+    if bitwise:
+        desired_op_classes.add(_OpClass.BITWISE)
+    if shift:
+        desired_op_classes.add(_OpClass.SHIFT)
+    if eq_comparison:
+        desired_op_classes.add(_OpClass.EQ_COMPARISON)
+    if rel_comparison:
+        desired_op_classes.add(_OpClass.REL_COMPARISON)
+
+    def wrap(cls):
+        from pytools.codegen import CodeGenerator
+        gen = CodeGenerator()
+        gen("""
+            from numbers import Number
+            import numpy as np
+            from meshmode.array_context import ArrayContainer
+            from pytools.obj_array import obj_array_vectorize
+            import operator as op
+            """)
+        gen("")
+
+        def same_key(k1, k2):
+            assert k1 == k2
+            return k1
+
+        # {{{ unary operators
+
+        for dunder_name, base_op_str, op_cls in _UNARY_OP_AND_DUNDER:
+            if op_cls not in desired_op_classes:
+                continue
+
+            fname = f"{cls.__name__.lower()}_{dunder_name}"
+            init_args = cls._deserialize_init_arrays_code("arg1", {
+                    key_arg1: _format_unary_op_str(base_op_str, expr_arg1)
+                    for key_arg1, expr_arg1 in
+                    cls._serialize_init_arrays_code("arg1").items()
+                    })
+
+            gen(f"""
+                def {fname}(arg1):
+                    return cls({init_args})
+                cls.__{dunder_name}__ = {fname}""")
+            gen("")
+
+        # }}}
+
+        for dunder_name, base_op_str, reversible, op_cls in _BINARY_OP_AND_DUNDER:
+            if op_cls not in desired_op_classes:
+                continue
+
+            # {{{ "forward" binary operators
+
+            fname = f"{cls.__name__.lower()}_{dunder_name}"
+            zip_init_args = cls._deserialize_init_arrays_code("arg1", {
+                    same_key(key_arg1, key_arg2):
+                    _format_binary_op_str(base_op_str, expr_arg1, expr_arg2)
+                    for (key_arg1, expr_arg1), (key_arg2, expr_arg2) in zip(
+                        cls._serialize_init_arrays_code("arg1").items(),
+                        cls._serialize_init_arrays_code("arg2").items())
```

Comment:
The `_serialize_init_arrays_code`/`_deserialize_init_arrays_code` is kind of like a code-gen-time/static version of the container interface. (It's also a bit of a hairball.) I tried to convince myself that we should get rid of one of these, but I wasn't able to. They both seem to serve a useful purpose. Arithmetic is too slow when using the run-time version of container serialization (as evidenced by the state before this PR). At the same time, interfacing with code-gen-time serialization is kind of a massive pain, and so I suspect it's not worth it for most code.